File: rpc/rpc.js
rpc = (function() {
  function Handles() {
    this.handles = {};
    this.nextId = 0;
  }

  Handles.prototype.allocate = function(object) {
    var id = (this.nextId++).toString();
    this.handles[id] = object;
    return id;
  }

  Handles.prototype.get = function(key) {
    if (!this.handles.hasOwnProperty(key)) throw new Error("Non-existent handle accessed: " + key);
    return this.handles[key];
  }

  Handles.prototype.release = function(key) {
    if (!this.handles.hasOwnProperty(key)) throw new Error("Non-existent handle freed: " + key);
    delete this.handles[key];
  }

  function Handle(home, id, endpoint, port, type) {
    this.home = home; // TODO(sammccall): just use port for this, once they can be compared
    this.id = id;
    this.endpoint = endpoint;
    this.port = port;
    this.handleType = type;
  }
  Handle.prototype.handleType = null;
  Handle.prototype.release = function() {
    invoke(this.endpoint, this.port, '__release__', [this.id]);
  };

  function RemoteFunction(home, id, endpoint, port) {
    function result() {
      return invoke(result.handle.endpoint, result.handle.port,
          '__call__', [result, Array.prototype.slice.call(arguments)]);
    }
    result.handle = new Handle(home, id, endpoint, port, 'function');
    result.release = function() { result.handle.release(); };
    return result;
  }

  function RemoteStream(home, id, endpoint, port) {
    async.Stream.call(this);
    this.handle = new Handle(home, id, endpoint, port, 'stream');
    this.registered = false;
  }
  RemoteStream.prototype = Object.create(async.Stream.prototype);
  RemoteStream.prototype.onItem = function(callback) {
    registerRemoteStream(this);
    async.Stream.prototype.onItem.call(this, callback);
  };
  RemoteStream.prototype.onClose = function(callback) {
    registerRemoteStream(this);
    async.Stream.prototype.onClose.call(this, callback);
  };
  function registerRemoteStream(stream) {
    if (stream.registered) return;
    var completer = new async.StreamSource(stream);
    invoke(stream.handle.endpoint, stream.handle.port, '__subscribe__', [stream.handle, function(result) {
      (result.closed) ? completer.close() : completer.emit(result.value);
    }]);
    stream.registered = true;
  }

  function RemoteFuture(home, id, endpoint, port) {
    async.Future.call(this);
    this.handle = new Handle(home, id, endpoint, port, 'future');
    this.registered = false;
  }
  RemoteFuture.prototype = Object.create(async.Future.prototype);
  RemoteFuture.prototype.onComplete = function(callback) {
    registerRemoteFuture(this);
    async.Future.prototype.onComplete.call(this, callback);
  };
  function registerRemoteFuture(future) {
    if (future.registered) return;
    var completer = new async.Completer(future);
    invoke(future.handle.endpoint, future.handle.port, '__subscribe__', [future.handle, function(result) {
      var err = result.hasOwnProperty('exception');
      completer.complete(
          err ? undefined : result.value,
          err ? new Error(result.exception) : undefined);
    }]);
    future.registered = true;
  }

  function Endpoint(name, methods) {
    this.id = generateId();
    this.name = name;
    this.methods = methods;
    this.serializer = new serialize.Serializer();
    this.deserializer = new serialize.Deserializer();
    this.receive = new ReceivePortSync();
    this.handles = new Handles();
    var endpoint = this;
    this.receive.receive(function(data) {
      debug(endpoint, "received", data);
      var ret;
      try {
        var args = endpoint.deserializer.apply(data.args);
        debug(endpoint, "invoking", data.method, args);
        var result = invokeHandler(endpoint, data.method, args);
        debug(endpoint, data.method, "returned", result);
        ret = {value: endpoint.serializer.apply(result)};
      } catch (e) {
        debug(endpoint, data.method, "threw", e, e.stack);
        ret = {exception: e.toString()};
      }
      debug(endpoint, "returning", ret);
      return ret;
    });
    function serializeHandle(handle) {
      if (!(handle instanceof Handle)) throw new Error("Not a handle");
      return ({
        $type: 'handle',
        id: handle.id,
        handleType: handle.handleType,
        home: handle.home,
        port: handle.port,
      })
    }
    this.serializer.register(function(x) { return x instanceof Function; }, function(func) {
      return serializeHandle(func.hasOwnProperty('handle') ? func.handle : endpoint.handle(func));
    });
    this.serializer.register(function(x) { return x instanceof Handle; }, serializeHandle);
    this.serializer.register(function(x) { return x instanceof async.Stream; }, function(stream) {
      if (stream instanceof RemoteStream) return serializeHandle(stream.handle);
      var handle = endpoint.handle(stream);
      stream.onClose(function() {
        endpoint.handles.release(handle.id);
      });
      return serializeHandle(handle);
    });
    this.serializer.register(function(x) { return x instanceof async.Future; }, function(future) {
      if (future instanceof RemoteFuture) return serializeHandle(future.handle);
      var handle = endpoint.handle(future);
      future.onComplete(function() {
        endpoint.handles.release(handle.id);
      });
      return serializeHandle(handle);
    });
    function typed(name) {
      return function(obj) { return (typeof obj == 'object') && (obj != null) && (obj.$type == name); };
    }
    this.deserializer.register(typed('handle'), function(obj) {
      if (obj.home == endpoint.id) return endpoint.handles.get(obj.id);
      var type = getHandleTypeByName(obj.handleType);
      return new type(obj.home, obj.id, endpoint, obj.port);
    });
  }

  function getHandleTypeByName(name) {
    return (name == 'stream') ? RemoteStream
        : (name == 'future') ? RemoteFuture
        : (name == 'function') ? RemoteFunction
        : Handle;
  }

  function getHandleType(obj) {
    return (obj instanceof async.Stream) ? 'stream'
        : (obj instanceof async.Future) ? 'future'
        : (obj instanceof Function) ? 'function'
        : null;
  }

  Endpoint.prototype.handle = function(object) {
    var id = this.handles.allocate(object);
    var type = getHandleType(object);
    return new Handle(this.id, id, this, this.receive.toSendPort(), type);
  }

  function Service(name, methodTable) {
    Endpoint.call(this, name, methodTable);
  }
  Service.prototype = Object.create(Endpoint.prototype);

  function invokeHandler(endpoint, name, args) {
    if (name == '__index__') return Object.keys(endpoint.methods);
    if (name == '__call__') return args[0].apply(null, args[1]);
    if (name == '__release__') return endpoint.handles.release(args[0]);
    if (name == '__subscribe__') {
      var target = args[0];
      var callback = args[1];
      if (target instanceof async.Stream) {
        target.onItem(function(x) { callback({value:x}); });
        target.onClose(function() { callback({closed:true}); callback.release(); });
      } else if (target instanceof async.Future) {
        target.onComplete(function(x, err) { 
          callback((err == null) ? {value:x} : {exception:err.toString()});
          callback.release();
        });
      } else {
        throw new Error("Tried to subscribe to something not a Stream or a Future");
      }
      return null;
    }
    if (endpoint.methods.hasOwnProperty(name)) return endpoint.methods[name].apply(null, args);
    throw new Error("No such method " + endpoint.name + "." + name);
  }

  Service.prototype.expose = function(name) {
    if (name == null) name = this.name;
    window.registerPort(name, this.receive.toSendPort());
  };

  function Client(name) {
    Endpoint.call(this, name, {});
    this.send = null;
  }
  Client.prototype = Object.create(Endpoint.prototype);

  function invoke(endpoint, port, method, args) {
    var args = endpoint.serializer.apply(args);
    var result = port.callSync({method:method, args:args});
    if (result.hasOwnProperty('exception')) throw new Error(result.exception);
    return endpoint.deserializer.apply(result.value);
  }

  Client.prototype.method = function(method) {
    var client = this;
    return function() {
      if (client.send == null) throw new Error("Not connected!");
      return invoke(client, client.send, method, Array.prototype.slice.call(arguments));
    }
  };

  Client.prototype.connect = function(name) {
    if (name == null) name = this.name;
    this.send = window.lookupPort(name);
    try {
      var methods = invoke(this, this.send, '__index__', []);
      for (var i = 0; i < methods.length; i++) {
        var method = methods[i];
        if (method in this) continue;
        this[method] = this.method(method);
      }
    } catch (e) {
      console.error("Failed to call __index__", e, e.stack);
    }
    return this;
  };

  function generateId() {
    return Math.floor(Math.random() * 1000000000).toString();
  }

  function debug() {
    var args = Array.prototype.slice.call(arguments);
    if (window.localStorage['debug_js'] == 'true') console.log.apply(console, args);
  }

  return {
    Service: Service,
    Client: Client,
    Handle: Handle,
    RemoteFunction: RemoteFunction,
  };
})();

File: rpc/rpc.dart
#library("rpc");
#import("dart:json");
//#import("dart:isolate", prefix: "isolate");
#import("dart:html");


class HandleTable {
  Map handles;
  HandleTable() {
    handles = new Map<String, Object>();
  }
  String nextId() {
    var id = (Math.random() * 1000000).toInt().toString();
    assert(!handles.containsKey(id));
    return id;
  }
  String registerObject(o) {
    var id = nextId();
    handles[id] = o;
    return id;
  }
  getObject(String id) {
    assert(handles.containsKey(id));
    return handles[id];
  }
  releaseObject(String id) {
    assert(handles.containsKey(id));
    handles.remove(id);
  }
}

class RemoteObject {
  String handleType;
  String id;
  String home;
  SendPortSync port;
  RPC rpc;
  
  RemoteObject(this.handleType, this.id, this.home, this.port, this.rpc) {
  }
  
  factory RemoteObject.fromJSON(Map json, RPC rpc) {
    assert(json["\$type"] == "handle");
    var handleType = json["handleType"];
    assert(handleType != null);
    var port = json["port"];
    if (handleType == 'function') {
      return new RemoteFunction(json["id"], json["home"], port, rpc);
    } else if (handleType == null) {
      return new RemoteObject(null, json["id"], json["home"], port, rpc);
    } else {
      throw "unsupported remote object type '$handleType'";
    }
  }

  Map serialize() {
    var serializedFunc = { "\$type": "handle",
             "handleType": handleType,
             "id": id,
             "port": port,
             "home": home };
    return serializedFunc;
  }
}

class RemoteFunction extends RemoteObject {
  RemoteFunction(String id, String home, SendPortSync port, RPC rpc)
    : super('function', id, home, port, rpc) {
  }
  
  String toString() =>
    "'RemoteFunction: {handleType: '$handleType', id: '$id'"
    ", home: '$home', port: '$port'}'";
  
  invoke(List args) {
    assert(port != null);
    assert(port is SendPortSync);
    return rpc.send(port, "__call__", [this, args]);
  }
}


class RPC {
  String _portName;
  String _homeId;
  ReceivePortSync _recvPort;
  Map<String, Function> _methods;
  HandleTable _handles;
    
  RPC.Server(String portName, Map methods) {
    _handles = new HandleTable();
    _portName = portName;
    _homeId = makeHomeId();
    _recvPort = new ReceivePortSync();
    _recvPort.receive(receiveCallback);
    window.registerPort(portName, _recvPort.toSendPort());
    _methods = methods;
  }
  
  RPC.Client(String portName) {
    _handles = new HandleTable();
    _portName = portName;
    _homeId = makeHomeId();
    _methods = null;
    _recvPort = new ReceivePortSync();
    _recvPort.receive(receiveCallback);
    // Debugging stuff. Check whether the server responds to the __index__
    // rpc call.
    var port = window.lookupPort(portName);
    assert(port != null);
    var res = port.callSync({"method":"__index__", "args":[]});
    debug(res);
  }

  String makeHomeId() => (Math.random() * 1000000).toInt().toString();

  SendPortSync mySendPort() {
    assert(_recvPort != null);
    return _recvPort.toSendPort();
  }

  serialize(obj) {
    if (obj is RemoteObject) {
      return obj.serialize();
    } else if (obj is Function) {
      var handleId = _handles.registerObject(obj);
      return { 
        "\$type": "handle",
        "handleType": "function",
        "id": handleId,
        "port": mySendPort(),
        "home": _homeId };
    } else if (obj is List) {
      var newList = new List(obj.length);
      for (int i = 0; i < obj.length; i++) {
        newList[i] = serialize(obj[i]);
      }
      return newList;
    } else if (obj is Map) {
      var newMap = new Map();
      obj.forEach((k, v) {
          newMap[serialize(k)] = serialize(v);
      });
      return newMap;
    } else {
      return obj;
    }
  }
  
  deserialize(obj) {
    if (obj is List) {
      for (int i = 0; i < obj.length; i++) {
        // We can be destructive, no need to copy.
        obj[i] = deserialize(obj[i]);
      }
    } else if ((obj is Map) && (obj["\$type"] == "handle")) {
      if (obj["home"] == _homeId) {
        assert(obj["id"] != null);
        obj = _handles.getObject(obj["id"]);
      } else {
        obj = new RemoteObject.fromJSON(obj, this);
      }
    }
    return obj;
  }

  apply(Function f, List args) {
    switch (args.length) {
      case 0: return f();
      case 1: return f(args[0]);
      case 2: return f(args[0], args[1]);
      case 3: return f(args[0], args[1], args[2]);
      default: throw "too many arguments!";
    }
  }

  receiveCallback(String msg) {
    debug("receiveCallback: $msg");
    String methodName = msg["method"];
    assert(methodName != null);
    List serializedArgs = msg["args"];
    assert(serializedArgs != null);
    List args = deserialize(serializedArgs);
    //debug("receiveCallback: deserialized args: $args");
    assert(args != null);
    if (methodName == "__index__") {
      return {"value": _methods.getKeys()};
    } else if (methodName == "__release__") {
      var id = args[0];
      var obj = _handles.getObject(id);
      //debug("request to release handle '$id' ($obj)");
      assert(id is String);
      assert(obj != null);
      _handles.releaseObject(id);
      return {"value": null};
    }
    Function m;
    if (methodName == "__call__") {
      assert(args.length == 2);
      m = args[0];     // First argument to rpc call is closure.
      args = args[1];  // Arguments to closure.
      //debug("request to call $m");
    } else {
      m = _methods[methodName];
      if (m == null) {
        return {"exception":"unknown rpc request '$methodName'"};
      }
    }
    assert(m is Function);
    try {
      var r = apply(m, args);
      r = serialize(r);
      return {"value": r};
    } catch(var e) {
      return {"exception": e.toString() };
    }
  }
  
  send(SendPortSync port, String methodName, List args) {
    debug("args before serialize: $args");
    var serializedArgs = serialize(args);
    var msg = {"method": methodName, "args": serializedArgs };
    debug("sending: $msg");
    var res = port.callSync(msg);
    debug("received: $res");
    if (res.containsKey("value")) {
      return deserialize(res["value"]);
    } else if (res["exception"] != null) {
      debug("received exception: ${res["exception"]}");
      throw res["exception"];
    } else {
      throw "Illegal rpc response format.";
    }
  }
  
  noSuchMethod(String methodName, List args) {
    debug(">>> noSuchMethod '$methodName' args: $args");
    var port = window.lookupPort(_portName);
    assert(port != null);
    return send(port, methodName, args);
  }
}

debug(message) {
  if (window.localStorage['debug_dart'] == 'true') print(message);
}

class CalcServer {
  RPC _rpc;
  
  generatePi(Map func_handle) {
    debug("generate_pi called with $func_handle");
    func_handle.invoke([4]);
    func_handle.invoke([2]);
    return null;
  }
  
  CalcServer() {
    var calcMethods = {
      "multiply": (a,b) => a * b,
      "add":      (a,b) => a + b,
      "subtract": (a,b) => a - b,
      "divide":   (a,b) => a / b,
      "generate_pi": (fh) => generatePi(fh)
    };
    _rpc = new RPC.Server("dart-calculator", calcMethods);
  }
}

testClient() {
  void handlePiDigit(digit) {
    debug("getting pi digit: $digit");
  }
  
  var jscalc= new RPC.Client("js-calculator");
  var sum = jscalc.add(10, 4);
  Expect.equals(14, sum);
  debug("js-calculator result: $sum");
  jscalc.generate_pi(handlePiDigit);
  try {
    var trash = jscalc.foo(100);
    Expect.fail("Error: failed to throw exception for non-existing method.");
  } catch (var e) {
  }
}

main() {
  testClient();
  var f = new CalcServer();
}
